blockindex/liquidstaking_indexer.go

```diff
@@ -0,0 +1,958 @@
+	for id, bi := range s.cache.getAllBucketInfo() {
+		bt := s.cache.mustGetBucketType(bi.TypeIndex)
+		vb, err := s.convertToVoteBucket(id, bi, bt)
+		if err != nil {
+			return nil, err
+		}
+		vbs = append(vbs, vb)
+	}
+	return vbs, nil
+}
+
+// Bucket returns the bucket
+func (s *liquidStakingIndexer) Bucket(id uint64) (*Bucket, error) {
+	bi, ok := s.cache.getBucketInfo(id)
+	if !ok {
+		return nil, errors.Wrapf(ErrBucketInfoNotExist, "id %d", id)
+	}
+	bt := s.cache.mustGetBucketType(bi.TypeIndex)
+	vb, err := s.convertToVoteBucket(id, bi, bt)
+	if err != nil {
+		return nil, err
+	}
+	return vb, nil
+}
+
+// BucketsByIndices returns the buckets by indices
+func (s *liquidStakingIndexer) BucketsByIndices(indices []uint64) ([]*Bucket, error) {
+	vbs := make([]*Bucket, 0, len(indices))
+	for _, id := range indices {
+		bi, ok := s.cache.getBucketInfo(id)
+		if !ok {
+			return nil, errors.Wrapf(ErrBucketInfoNotExist, "id %d", id)
+		}
+		bt := s.cache.mustGetBucketType(bi.TypeIndex)
+		vb, err := s.convertToVoteBucket(id, bi, bt)
+		if err != nil {
+			return nil, err
+		}
+		vbs = append(vbs, vb)
+	}
+	return vbs, nil
+}
+
+func (s *liquidStakingIndexer) TotalBucketCount() uint64 {
+	return s.cache.getTotalBucketCount()
+}
+
+func (s *liquidStakingIndexer) ActiveBucketTypes() (map[uint64]*BucketType, error) {
+	return s.cache.getActiveBucketType(), nil
+}
+
+func (s *liquidStakingIndexer) handleEvent(ctx context.Context, dirty *liquidStakingDirty, blk *block.Block, act *action.SealedEnvelope, log *action.Log) error {
+	// get event abi
+	abiEvent, err := _liquidStakingInterface.EventByID(common.Hash(log.Topics[0]))
+	if err != nil {
+		return errors.Wrapf(err, "get event abi from topic %v failed", log.Topics[0])
+	}
+
+	// unpack event data
+	event, err := unpackEventParam(abiEvent, log)
+	if err != nil {
+		return err
+	}
+
+	// handle different kinds of event
+	switch abiEvent.Name {
+	case "BucketTypeActivated":
+		return s.handleBucketTypeActivatedEvent(dirty, event, blk.Height())
+	case "BucketTypeDeactivated":
+		return s.handleBucketTypeDeactivatedEvent(dirty, event, blk.Height())
+	case "Staked":
+		return s.handleStakedEvent(dirty, event, blk.Height())
+	case "Locked":
+		return s.handleLockedEvent(dirty, event)
+	case "Unlocked":
+		return s.handleUnlockedEvent(dirty, event, blk.Height())
+	case "Unstaked":
+		return s.handleUnstakedEvent(dirty, event, blk.Height())
+	case "Merged":
+		return s.handleMergedEvent(dirty, event)
+	case "DurationExtended":
+		return s.handleDurationExtendedEvent(dirty, event)
+	case "AmountIncreased":
+		return s.handleAmountIncreasedEvent(dirty, event)
+	case "DelegateChanged":
+		return s.handleDelegateChangedEvent(dirty, event)
+	case "Withdrawal":
+		return s.handleWithdrawalEvent(dirty, event)
+	case "Transfer":
+		return s.handleTransferEvent(dirty, event)
+	default:
+		return nil
+	}
+}
+
+func (s *liquidStakingIndexer) handleTransferEvent(dirty *liquidStakingDirty, event eventParam) error {
```

Comment:
not function of `linquidStakingIndexer`, but `liquidStakingDirty`. Same for all the `handleXXX` functions

Reply:
a good suggestion, making code more clearly